File: src/main/kotlin/com/isel/ps/secdash/model/repositories/Repository.kt
package com.isel.ps.secdash.model.repositories

import com.isel.ps.secdash.model.Owner
import com.isel.ps.secdash.model.Platform
import com.isel.ps.secdash.model.repositories.Repository.Visibility
import java.time.Instant

class Repository(
    val rid: Int,
    val name: String,
    val externalId: String,
    val platform: Platform,
    val owner: Owner,
    val htmlUrl: String,
    val description: String,
    val issuesCount: Int,
    val createdAt: Instant,
    val updatedAt: Instant,
    val forksCount: Int,
    val visibility: Visibility
    ) {
    enum class Visibility {
        PUBLIC, PRIVATE, INTERNAL
    }
}



data class RepositorySqlDto(
    val rid: Int,
    val name: String,
    val externalId: String,
    val platform: Platform,
    val ownerId: Int,
    val htmlUrl: String,
    val description: String,
    val issuesCount: Int,
    val createdAt: Instant,
    val updatedAt: Instant,
    val forksCount: Int,
    val visibility: Visibility
) {
    fun toDomainRepository(owner: Owner) = Repository(
        rid = rid,
        name = name,
        externalId = externalId,
        platform = platform,
        owner = owner,
        htmlUrl = htmlUrl,
        description = description,
        issuesCount = issuesCount,
        createdAt = createdAt,
        updatedAt = updatedAt,
        forksCount = forksCount,
        visibility = visibility,
    )
}

data class RepositoryWithOwnerSqlDto(
    val rid: Int,
    val name: String,
    val externalId: String,
    val platform: Platform,
    val ownerId: Int,
    val htmlUrl: String,
    val description: String?,
    val issuesCount: Int,
    val createdAt: Instant,
    val updatedAt: Instant,
    val forksCount: Int,
    val visibility: Visibility,
    val oid: Int,
    val oExternalId: String,
    val oName: String,
    val url: String,
    val avatarUrl: String?,
    val oPlatform: Platform,
) {
    fun toDomain() = Repository(
        rid = rid,
        name = name,
        externalId = externalId,
        platform = platform,
        owner = Owner(
            oid = oid,
            externalId = oExternalId,
            name = oName,
            url = url,
            avatarUrl = avatarUrl,
            platform = oPlatform,
        ),
        htmlUrl = htmlUrl,
        description = description ?: "",
        issuesCount = issuesCount,
        createdAt = createdAt,
        updatedAt = updatedAt,
        forksCount = forksCount,
        visibility = visibility,
    )
}

File: frontend/src/model/repository/repository.ts
export type Platform = "GITHUB" | "GITLAB"

export type Visibility = "PUBLIC" | "PRIVATE" | "INTERNAL"

export type Owner = {
    oid: number,
    externalId: string,
    name: string,
    url: string,
    avatarUrl: string | null,
    platform: Platform,
}

export type Repository = {
    rid: number,
    name: string,
    externalId: string,
    platform: Platform,
    owner: Owner,
    htmlUrl: string,
    description: string,
    issuesCount: number,
    createdAt: string,
    updatedAt: string,
    forksCount: number,
    visibility: Visibility,
}

export type CountsBySeverity = {
    critical: number
    high: number
    medium: number
    low: number
    unknown: number
}

export type VulnerabilityStats = {
    open: number
    fixed: number
    dismissed: number
    countsBySeverity: CountsBySeverity
}

export type SastStats = {
    open: number
    fixed: number
    dismissed: number
    countsBySeverity: CountsBySeverity
}

export type RepoStats = {
    vulnerabilityStats: VulnerabilityStats
    sastStats: SastStats
}

export type DailySastCount = {
    date: string
    count: number
    countsBySeverity: CountsBySeverity
}

export type DailyVulnerabilityCount = {
    date: string
    count: number
    countsBySeverity: CountsBySeverity
}

export type DailySastCountList = DailySastCount[]

export type DailyVulnerabilityCountList = DailyVulnerabilityCount[]

export type HistoryStats = {
    sastList: DailySastCountList
    vulnList: DailyVulnerabilityCountList
}
